streams/hash/others/hash_functions/gost/gost.cpp

```diff
@@ -0,0 +1,879 @@
+                   (v[1] >> 16) ^ (v[1] & 0xffff0000) ^ (v[2] << 16) ^
+                   (v[3] >> 16) ^ (v[4] << 16) ^ (v[5] >> 16) ^ v[5] ^
+                   (v[6] >> 16) ^ (v[7] << 16) ^ (v[7] >> 16) ^ (v[7] & 0xffff);
+    ctx->hash[1] = (v[0] << 16) ^ (v[0] >> 16) ^ (v[0] & 0xffff0000) ^
+                   (v[1] & 0xffff) ^ v[2] ^ (v[2] >> 16) ^ (v[3] << 16) ^
+                   (v[4] >> 16) ^ (v[5] << 16) ^ (v[6] << 16) ^ v[6] ^
+                   (v[7] & 0xffff0000) ^ (v[7] >> 16);
+    ctx->hash[2] = (v[0] & 0xffff) ^ (v[0] << 16) ^ (v[1] << 16) ^
+                   (v[1] >> 16) ^ (v[1] & 0xffff0000) ^ (v[2] << 16) ^ (v[3] >> 16) ^
+                   v[3] ^ (v[4] << 16) ^ (v[5] >> 16) ^ v[6] ^ (v[6] >> 16) ^
+                   (v[7] & 0xffff) ^ (v[7] << 16) ^ (v[7] >> 16);
+    ctx->hash[3] = (v[0] << 16) ^ (v[0] >> 16) ^ (v[0] & 0xffff0000) ^
+                   (v[1] & 0xffff0000) ^ (v[1] >> 16) ^ (v[2] << 16) ^
+                   (v[2] >> 16) ^ v[2] ^ (v[3] << 16) ^ (v[4] >> 16) ^ v[4] ^
+                   (v[5] << 16) ^ (v[6] << 16) ^ (v[7] & 0xffff) ^ (v[7] >> 16);
+    ctx->hash[4] = (v[0] >> 16) ^ (v[1] << 16) ^ v[1] ^ (v[2] >> 16) ^ v[2] ^
+                   (v[3] << 16) ^ (v[3] >> 16) ^ v[3] ^ (v[4] << 16) ^
+                   (v[5] >> 16) ^ v[5] ^ (v[6] << 16) ^ (v[6] >> 16) ^ (v[7] << 16);
+    ctx->hash[5] = (v[0] << 16) ^ (v[0] & 0xffff0000) ^ (v[1] << 16) ^
+                   (v[1] >> 16) ^ (v[1] & 0xffff0000) ^ (v[2] << 16) ^ v[2] ^
+                   (v[3] >> 16) ^ v[3] ^ (v[4] << 16) ^ (v[4] >> 16) ^ v[4] ^
+                   (v[5] << 16) ^ (v[6] << 16) ^ (v[6] >> 16) ^ v[6] ^
+                   (v[7] << 16) ^ (v[7] >> 16) ^ (v[7] & 0xffff0000);
+    ctx->hash[6] = v[0] ^ v[2] ^ (v[2] >> 16) ^ v[3] ^ (v[3] << 16) ^ v[4] ^
+                   (v[4] >> 16) ^ (v[5] << 16) ^ (v[5] >> 16) ^ v[5] ^
+                   (v[6] << 16) ^ (v[6] >> 16) ^ v[6] ^ (v[7] << 16) ^ v[7];
+    ctx->hash[7] = v[0] ^ (v[0] >> 16) ^ (v[1] << 16) ^ (v[1] >> 16) ^
+                   (v[2] << 16) ^ (v[3] >> 16) ^ v[3] ^ (v[4] << 16) ^ v[4] ^
+                   (v[5] >> 16) ^ v[5] ^ (v[6] << 16) ^ (v[6] >> 16) ^ (v[7] << 16) ^ v[7];
+}
+
+/**
+ * This function calculates hash value by 256-bit blocks.
+ * It updates 256-bit check sum as follows:
+ *    *(uint256_t)(ctx->sum) += *(uint256_t*)block;
+ * and then updates intermediate hash value ctx->hash
+ * by calling rhash_gost_block_compress().
+ *
+ * @param ctx algorithm context
+ * @param block the 256-bit message block to process
+ */
+static void rhash_gost_compute_sum_and_hash(gost_ctx * ctx, const unsigned* block, unsigned nr)
+{
+#if IS_LITTLE_ENDIAN
+    # define block_le block
+# define LOAD_BLOCK_LE(i)
+#else
+    unsigned block_le[8]; /* tmp buffer for little endian number */
+# define LOAD_BLOCK_LE(i) (block_le[i] = le2me_32(block[i]))
+#endif
+
+    /* This optimization doesn't improve speed much,
+    * and saves too little memory, but it was fun to write! =)  */
```

Comment:
Who the fuck wrote this? :D

(Yes, I see it in original code https://github.com/rhash/RHash/blob/master/librhash/gost.c#L291)

Reply:
:D dont know.